src/sequential/sequential.cpp

```diff
@@ -0,0 +1,275 @@
+SequentialBucket::swap(SequentialBucket &other) {
+  LOG_TRACE("Enter");
+  KeyType tmp_key = this->key;
+  this->key = other.key;
+  other.key = tmp_key;
+
+  ValueType tmp_value = this->value;
+  this->value = other.value;
+  other.value = tmp_value;
+
+  HashCodeType tmp_hc = this->hashcode;
+  this->hashcode = other.hashcode;
+  other.hashcode = tmp_hc;
+
+  size_t tmp_offset = this->offset;
+  this->offset = other.offset;
+  other.offset = tmp_offset;
+}
+
+bool
+SequentialBucket::equal_key(const KeyType key, const HashCodeType hashcode) {
+  LOG_TRACE("Enter");
+  assert(!this->is_empty() && "comparing to empty bucket!");
+  // Assume equality between hashcodes is simpler (because it is fixed for any
+  // type of key)
+  return this->hashcode == hashcode && this->key == key;
+}
+
+void
+SequentialBucket::print(const size_t size) {
+  if (this->is_empty()) {
+    std::cout << "(empty)";
+  } else {
+    std::cout << "(" << this->hashcode << "=>" << this->hashcode % size <<
+        "+" << this->offset << ") " << this->key << ": " << this->value;
+  }
+}
+
+
+////////////////////////////////////////////////////////////////////////////////
+/// STATIC HELPER FUNCTIONS
+////////////////////////////////////////////////////////////////////////////////
+
+HashCodeType
+hash(const KeyType key) {
+  LOG_TRACE("Enter");
+  size_t k = static_cast<size_t>(key);
+  // I use the suffix '*ULL' to denote that the literal is an int64. I've had
+  // weird bugs in the past to do with literal conversion. I'm not sure the
+  // details. I only remember it was a right pain.
+  k = ((k >> 30) ^ k) * 0xbf58476d1ce4e5b9ULL;
+  k = ((k >> 27) ^ k) * 0x94d049bb133111ebULL;
+  k =  (k >> 31) ^ k;
+  return reinterpret_cast<HashCodeType>(k);
+}
+
+size_t
+get_home(const HashCodeType hashcode, const size_t size) {
+  LOG_TRACE("Enter");
+  size_t h = reinterpret_cast<HashCodeType>(hashcode);
+  return h % size;
+}
+
+/// @brief  Get real bucket index.
+size_t
+get_probe(const size_t home, const size_t offset, const size_t size) {
+  LOG_TRACE("Enter");
+  return (home + offset) % size;
+}
+
+std::pair<SearchStatus, size_t>
+get_wouldbe_offset(      std::vector<SequentialBucket> &tmp_buckets,
+                   const KeyType key,
+                   const HashCodeType hashcode,
+                   const size_t home) {
+  LOG_TRACE("Enter");
+  size_t i = 0;
+  size_t size = tmp_buckets.size();
+  for (i = 0; i < size; ++i) {
+    size_t real_index = get_probe(home, i, tmp_buckets.size());
+    SequentialBucket &bkt = tmp_buckets[real_index];
+    // If not found
+    if (bkt.is_empty()) {
+      // This is first, because equality on an empty bucket is not well defined.
+      return {FOUND_HOLE, i};
+    } else if (bkt.offset < i) { // This means that bkt belongs to a nearer home
+      return {FOUND_SWAP, i};
+    // If found
+    } else if (bkt.equal_key(key, hashcode)) {
+      return {FOUND_MATCH, i};
+    }
+  }
+  return {FOUND_NOHOLE, SIZE_MAX};
+}
+
+int
```

Comment:
Hmm, I don't love the semantics of this. The idiomatic way to error check (in C) is:

```c
int r = insert_without_resize(...);
assert(!r && "error in insert_without_resize");
```

or

```c
if (insert_without_resize(...)) {
    /* ...handle error... */
}
```

Reply:
I agree with the semantics of zero and and non-zero in the case that the return type is int. But afaik, int (or, even better, an enum) is generally used when you need to report multiple possible error conditions.

Otherwise, if you just have 2 unambiguous states, bool is enough. 

If you use bool, it reads more naturally to indicate the positive case with `true` and the negative case with `false`. I think it's a pretty common convention too, considering Linux does the same thing. They define [a function returning bool](https://github.com/chewitt/linux/blob/ffc253263a1375a65fa6c9f62a893e9767fbebfa/mm/slab.c#L1870), and [use it as I showed above](https://github.com/chewitt/linux/blob/ffc253263a1375a65fa6c9f62a893e9767fbebfa/mm/slab.c#L2029).

Reply:
I made an enum error type.

Reply:
Nice, enum is the best case scenario here I think.